sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala

```diff
@@ -125,6 +135,31 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
       isTemporary = isTemp)
   }
 
+  private def makeTable(ident: Seq[String]): Table = {
+    val plan = UnresolvedTableOrView(ident, "Catalog.listTables", true)
+    val node = sparkSession.sessionState.executePlan(plan).analyzed
+    node match {
+      case t: ResolvedTable =>
+        val isExternal = t.table.properties().getOrDefault("external", "false").equals("true")
+        new Table(
+          name = t.identifier.name(),
+          database = t.identifier.namespace().head,
+          description = t.table.properties().get("comment"),
+          tableType =
+            if (isExternal) CatalogTableType.EXTERNAL.name
+            else CatalogTableType.MANAGED.name,
+          isTemporary = false)
+      case v: ResolvedView =>
+        new Table(
```

Comment:
my major concern is this `Table` API. It needs to support n-part name but keep backward compatibility. I'm thinking about
```
class Table(
  val name: String,
  val qualifiers: Array[String],
  ...
) {
  def database: String = if (qualifiers.length == 2) qualifiers(1) else null
}
```

Reply:
We will hit similar problems in `listDatabases` and `listFunctions`

Reply:
Why `database` is null if `qualifier.length` != 2?

`qualifier` could contain only one string which is a database name?

Reply:
if `qualifier.length > 2`, it's a new code path and I don't know what to report as a "database name". NULL seems reasonable here.

Reply:
pushed a new commit. Can you take a look?

Reply:
BTW this idea triggered a compatibility check issue
```
spark-sql: Failed binary compatibility check against org.apache.spark:spark-sql_2.12:3.2.0! Found 1 potential problems (filtered 586)
[error]  * method this(java.lang.String,java.lang.String,java.lang.String,java.lang.String,Boolean)Unit in class org.apache.spark.sql.catalog.Table's type is different in current version, where it is (java.lang.String,Array[java.lang.String],java.lang.String,java.lang.String,Boolean)Unit instead of (java.lang.String,java.lang.String,java.lang.String,java.lang.String,Boolean)Unit
```
How do we usually solve such issues?

Reply:
add a new constructor to match the old class definition